packages/engine/lib/error/src/ext.rs

```diff
@@ -0,0 +1,140 @@
+        C: Context,
+    {
+        match self {
+            Ok(t) => Ok(t),
+            Err(error) => Err(Report::from(error).context(context)),
+        }
+    }
+
+    #[track_caller]
+    fn with_context<C, F>(self, context: F) -> Result<T, Report>
+    where
+        C: Context,
+        F: FnOnce() -> C,
+    {
+        match self {
+            Ok(t) => Ok(t),
+            Err(error) => Err(Report::from(error).context(context())),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl<T, E> ReportErrorKind<T> for Result<T, E>
+where
+    E: std::error::Error + Send + Sync + 'static,
+{
+    #[track_caller]
+    fn error_kind<K>(self, error_kind: K) -> Result<T, Report<K>>
+    where
+        K: ErrorKind,
+    {
+        match self {
+            Ok(t) => Ok(t),
+            Err(error) => Err(Report::from(error).error_kind(error_kind)),
+        }
+    }
+
+    #[track_caller]
+    fn with_error_kind<K, F>(self, error_kind: F) -> Result<T, Report<K>>
+    where
+        K: ErrorKind,
+        F: FnOnce() -> K,
+    {
+        match self {
+            Ok(t) => Ok(t),
+            Err(error) => Err(Report::from(error).error_kind(error_kind())),
+        }
+    }
+}
+
+impl<T, E> ReportContext<T> for Result<T, Report<E>> {
+    type ErrorKind = E;
+
+    #[track_caller]
+    fn context<C>(self, context: C) -> Self
+    where
+        C: Context,
+    {
+        match self {
+            Ok(t) => Ok(t),
+            Err(report) => Err(report.context(context)),
+        }
+    }
+
+    #[track_caller]
+    fn with_context<C, F>(self, context: F) -> Self
+    where
+        C: Context,
+        F: FnOnce() -> C,
+    {
+        match self {
+            Ok(t) => Ok(t),
+            Err(report) => Err(report.context(context())),
+        }
+    }
+}
+
+impl<T, E> ReportErrorKind<T> for Result<T, Report<E>> {
+    #[track_caller]
+    fn error_kind<K>(self, error_kind: K) -> Result<T, Report<K>>
+    where
+        K: ErrorKind,
+    {
+        match self {
+            Ok(t) => Ok(t),
+            Err(report) => Err(report.error_kind(error_kind)),
+        }
+    }
+
+    #[track_caller]
+    fn with_error_kind<K, F>(self, error_kind: F) -> Result<T, Report<K>>
+    where
+        K: ErrorKind,
+        F: FnOnce() -> K,
+    {
+        match self {
+            Ok(t) => Ok(t),
+            Err(report) => Err(report.error_kind(error_kind())),
+        }
+    }
+}
+
+impl<T> ReportContext<T> for Option<T> {
+    type ErrorKind = ();
+
+    #[track_caller]
+    fn context<C>(self, context: C) -> Result<T, Report>
+    where
+        C: Context,
+    {
+        match self {
+            Some(t) => Ok(t),
+            None => Err(Report::new(context)),
+        }
+    }
+
+    #[track_caller]
+    fn with_context<C, F>(self, context: F) -> Result<T, Report>
+    where
+        C: Context,
+        F: FnOnce() -> C,
+    {
+        match self {
+            Some(t) => Ok(t),
+            None => Err(Report::new(context())),
+        }
+    }
+}
```

Comment:
Snippet from the [`eyre` documentation](https://docs.rs/eyre/latest/eyre/index.html):
> ### Why Doesn't `Eyre` impl `WrapErr` for `Option`?
>
> `eyre` doesn't impl `WrapErr` for `Option` because `wrap_err` implies that you're creating a new error that saves the previous error as its `source`. Calling `wrap_err` on an `Option` is meaningless because there is no source error. `anyhow` avoids this issue by using a different mental model where you're adding "context" to an error, though this not a mental model for error handling that `eyre` agrees with.
>
> Instead, `eyre` encourages users to think of each error as distinct, where the previous error is the context being saved by the new error, which is backwards compared to `anyhow`'s model. In this model you're encouraged to use combinators provided by `std` for `Option` to convert an option to a `Result`

How do we want to proceed?

Reply:
Personally I like the reasoning in `eyre` and propose to remove our implementation for `Option`

Reply:
This way, `ReportContext` and `ReportErrorKind` could be unified.

Reply:
I agree. I don't view options as almost ever being errors in themselves, otherwise a Result would be used. It makes sense to me that you'd create a new error when you get an unexpected None type, rather than converting the None into an error.